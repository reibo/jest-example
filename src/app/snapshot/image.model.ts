import {SafeUrl} from '@angular/platform-browser';

export interface Image {
  readonly src: SafeUrl;
  readonly id: number;
  readonly cols: number;
  readonly rows: number;
}
