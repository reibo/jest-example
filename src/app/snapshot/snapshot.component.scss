mat-grid-tile {
  border: 2px solid gray;

  div {
    background-repeat: no-repeat;
    background-size: cover;
    height: 100%;
    width: 100%;
  }

}
