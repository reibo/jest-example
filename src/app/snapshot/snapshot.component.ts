import {Component, Input} from '@angular/core';
import {Image} from './image.model';
import {DomSanitizer} from '@angular/platform-browser';

const IMAGE_SOURCE = 'https://picsum.photos/600';

@Component({
  selector: 'app-snapshot',
  template: `
    <mat-grid-list [cols]="maxCols" rowHeight="100px">
      <mat-grid-tile
        *ngFor="let image of images"
        [colspan]="image.cols"
        [rowspan]="image.rows">
        <div
          [style.background-image]="image.src"></div>
      </mat-grid-tile>
    </mat-grid-list>
  `,
  styleUrls: ['./snapshot.component.scss']
})
export class SnapshotComponent {
  maxCols = 6;

  images: Image[];

  constructor(private sanitizer: DomSanitizer) {}

  @Input() set totalImages(totalImages: number) {
    this.maxCols = totalImages % 6 || 6;
    const images: Image[] = [];
    for (let i = 0; i < totalImages; i++) {
      images.push({
        id: i,
        src: this.sanitizer.bypassSecurityTrustStyle(this.generateRandomUrl(i)),
        rows: this.getRows(i, totalImages),
        cols: this.getCols(i, totalImages)
      });
    }
    this.images = images;
  }

  private getRows(id: number, totalElements: number) {
    return (totalElements - id) % this.maxCols || 1;
  }


  private getCols(id: number, totalElements: number) {
    return Math.floor((id % this.maxCols) / 3 + 1);
  }

  private generateRandomUrl(id: number) {
    return `url("${IMAGE_SOURCE}?random=${id})`;
  }
}
