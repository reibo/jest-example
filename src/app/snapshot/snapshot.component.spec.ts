import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { SnapshotComponent } from './snapshot.component';
import {MatGridListModule} from '@angular/material';

describe('SnapshotComponent', () => {
  let component: SnapshotComponent;
  let fixture: ComponentFixture<SnapshotComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ SnapshotComponent ],
      imports: [MatGridListModule]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SnapshotComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('we have 10 images to visualise', () => {
    it('should render 10 images in our template', () => {
      component.totalImages = 10;
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });
});
