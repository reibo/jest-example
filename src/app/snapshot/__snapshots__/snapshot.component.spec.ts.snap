// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapshotComponent we have 10 images to visualise should render 10 images in our template 1`] = `
<app-snapshot
  images={[Function Array]}
  maxCols={[Function Number]}
  sanitizer={[Function DomSanitizerImpl]}
>
  <mat-grid-list
    class="mat-grid-list"
    ng-reflect-cols="4"
    ng-reflect-row-height="100px"
    rowheight="100px"
  >
    <div>
      
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="2"
        style="left: 0px; top: 0px;"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="1"
        style="top: 0px;"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="1"
        style="top: 0px;"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="2"
        ng-reflect-rowspan="3"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="2"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="1"
        style="left: 0px;"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="1"
        style="left: 0px;"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="2"
        ng-reflect-rowspan="3"
        style="left: 0px;"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="2"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
      <mat-grid-tile
        class="mat-grid-tile"
        ng-reflect-colspan="1"
        ng-reflect-rowspan="1"
      >
        <figure
          class="mat-figure"
        >
          <div />
        </figure>
      </mat-grid-tile>
    </div>
  </mat-grid-list>
</app-snapshot>
`;
